flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractHandler.java

```diff
@@ -160,28 +160,46 @@ protected void respondAsLeader(ChannelHandlerContext ctx, RoutedRequest routedRe
 			final FileUploads finalUploadedFiles = uploadedFiles;
 			requestProcessingFuture
 				.whenComplete((Void ignored, Throwable throwable) -> {
-					inFlightRequestTracker.deregisterRequest();
-					cleanupFileUploads(finalUploadedFiles);
+					if (throwable != null) {
+						handleException(ExceptionUtils.stripCompletionException(throwable), ctx, httpRequest);
+					}
+					finalizeRequestProcessing(finalUploadedFiles);
```

Comment:
We are slightly changing semantics here, because before we finalized the request processing after the error message has been sent. Now we send the error message and without waiting for the completion we finalize the request right away. One could change it to `requestProcessingFuture.exceptionally(handleException).thenCompose(Function.identity).whenComplete()`. But I think it should also be ok to finalize the request right away since we only send an error message and whether the error message arrives or the client connection is terminated via a `ConnectionException` should not matter too much.

Reply:
I've made changes to ensure that we finalize the request after the error response was sent.